resolver.go

```diff
@@ -308,12 +308,97 @@ func setFields(ctx *hcontext, req *http.Request, input reflect.Value, t reflect.
 			inField.Set(reflect.ValueOf(parsed))
 		}
 	}
+}
+
+// A smart join for JSONPath
+func pathJoin(prefix string, parts ...string) string {
+	joined := prefix
+	if joined != "" {
+		joined += "."
+	}
+	return joined + strings.Join(parts, ".")
+}
+
+// ctxLocationWrapper wraps a context so that the error detail `location` field
+// gets sets appropriately for resolver errors. I.e. the resolver doesn't know
+// when it runs whether it is the body or deeply nested within the body of an
+// incoming request. We prefix it so the errors make sense to the end-user.
+type ctxLocationWrapper struct {
+	*hcontext
+	location string
+}
+
+func (c ctxLocationWrapper) AddError(err error) {
+	if e, ok := err.(*ErrorDetail); ok {
+		e.Location = pathJoin(c.location, e.Location)
+	}
+
+	c.hcontext.AddError(err)
+}
+
+// resolveFields recursively crawls the input struct and calls Resolve on
+// any structs it finds as fields, within slices, and as values in maps. This
+// should be called *after* all other fields are set so the resolver code can
+// use their values. It processes depth-first so structs have access to the
+// resolved fields of any contained structs when their resolver runs.
+func resolveFields(ctx *hcontext, path string, input reflect.Value) {
+	if input.Kind() == reflect.Ptr {
+		input = input.Elem()
```

Comment:
I wish I could say more than "code smell" but some part of me wants to see this turn into a recursive call on `input.Elem()` rather than just shifting to what `input` points. I wish I had a more definitive response than the vague feeling that there's a pathological case where this goes wrong that I'm not clever enough to articulate. I'm okay to approve it as-is but I wanted to highlight it as a discussion point.

Reply:
Great catch actually. Doing this recursively means we'll better support pointers of pointers, though they are exceedingly rare in Go.